geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala

```diff
@@ -411,24 +417,21 @@ object DeltaWriter extends StrictLogging {
 
     val dictionaries = mergedDictionaries.dictionaries
 
-    // gets the attribute we're sorting by from the i-th feature in the vector
-    val getSortAttribute: (ArrowAttributeReader, scala.collection.Map[Integer, Integer], Int) => AnyRef = {
-      if (dictionaries.contains(sortBy)) {
-        // since we've sorted the dictionaries, we can just compare the encoded index values
-        (reader, mappings, i) => mappings(reader.asInstanceOf[ArrowDictionaryReader].getEncoded(i))
-      } else {
-        (reader, _, i) => reader.apply(i)
-      }
-    }
-
     val result = SimpleFeatureVector.create(sft, dictionaries, encoding)
     val unloader = new RecordBatchUnloader(result)
 
     logger.trace(s"merging sorted deltas - read schema: ${result.underlying.getField}")
 
-    // builder for our merge array - (vector, reader for sort values, transfers to result, dictionary mappings)
-    val mergeBuilder = Array.newBuilder[(SimpleFeatureVector, ArrowAttributeReader, Seq[(Int, Int) => Unit], scala.collection.Map[Integer, Integer])]
-    mergeBuilder.sizeHint(threadedBatches.foldLeft(0)((sum, a) => sum + a.length))
+    // we do a merge sort of each batch
+    // queue sorted by current value in each batch
+    val queue = {
+      val ordering = if (reverse) { BatchMergerOrdering.reverse } else { BatchMergerOrdering }
+      new PriorityQueue[BatchMerger[Any]](ordering)
+    }
+
+    // track our open vectors to close later
+    val cleanup = ArrayBuffer.empty[SimpleFeatureVector]
+    cleanup.sizeHint(threadedBatches.foldLeft(0)((sum, a) => sum + a.length))
```

Comment:
threadedBatches.map(_.length).sum too pedestrian?:)

Reply:
no point in creating an intermediate collection :)

Reply:
Fair enough.  

At the rate we're going, we're gonna start writing this stuff in JVM bytecode directly.;)